perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_group_attribute_def_def_uniqueID.java

```diff
@@ -0,0 +1,73 @@
+package cz.metacentrum.perun.core.impl.modules.attributes;
+
+import cz.metacentrum.perun.core.api.Attribute;
+import cz.metacentrum.perun.core.api.AttributeDefinition;
+import cz.metacentrum.perun.core.api.AttributesManager;
+import cz.metacentrum.perun.core.api.BeansUtils;
+import cz.metacentrum.perun.core.api.Group;
+import cz.metacentrum.perun.core.api.exceptions.InternalErrorException;
+import cz.metacentrum.perun.core.api.exceptions.WrongAttributeAssignmentException;
+import cz.metacentrum.perun.core.impl.PerunSessionImpl;
+import cz.metacentrum.perun.core.implApi.modules.attributes.GroupAttributesModuleAbstract;
+import cz.metacentrum.perun.core.implApi.modules.attributes.GroupAttributesModuleImplApi;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+/**
+ * Class for checking and filling unique IDs for groups in the Perun.
+ * This unique ID is unpredictable and is usable for identifying of groups in other systems.
+ *
+ * @author  Michal Stava stavamichal@gmail.com
+ */
+public class urn_perun_group_attribute_def_def_uniqueID  extends GroupAttributesModuleAbstract implements GroupAttributesModuleImplApi {
+
+	@Override
+	public Attribute fillAttribute(PerunSessionImpl session, Group group, AttributeDefinition attribute) throws InternalErrorException, WrongAttributeAssignmentException {
+		Attribute filledAttribute = new Attribute(attribute);
+		filledAttribute.setValue(sha1HashCount(group).toString());
+		return filledAttribute;
+	}
+
+	/**
+	 * Generate unique ID as hexadecimal string representation of SHA1 digest from group ID.
+	 * Input is salted per Perun instance. Effective resulting string consist of [0-9a-f] characters.
+	 *
+	 * It is used to generate unique IDs for Perun groups.
+	 *
+	 * @param group Group to generate ID for
+	 * @return Builder to get string ID
+	 * @throws InternalErrorException When generation fails
+	 */
+	protected StringBuilder sha1HashCount(Group group) throws InternalErrorException {
+		try {
+			String salt = BeansUtils.getCoreConfig().getInstanceId();
+			MessageDigest mDigest = MessageDigest.getInstance("SHA1");
+			// counts sha1hash and converts output to hex
+			int length = 4+salt.getBytes(StandardCharsets.UTF_8).length;
+			byte[] result = mDigest.digest(ByteBuffer.allocate(length).putInt(group.getId()).put(salt.getBytes(StandardCharsets.UTF_8)).array());
```

Comment:
NIT: This is really difficult to read. Something like this would be much easier:
```
byte[] result = mDigest.digest(ByteBuffer
				.allocate(length)
				.putInt(group.getId())
				.put(salt.getBytes(StandardCharsets.UTF_8))
				.array());
```

Reply:
Changed.